docs/concepts.md

```diff
@@ -0,0 +1,320 @@
+
+In System Dynamics there's no "movement" concept and generally Entities are
+represented by a count or summation in a Stock. A Stock in System Dynamics is
+connected to other Stocks via Flows. Like Stocks these are functions of time
+elapsed, but they are derivatives, not integrals: rates of change, rather than
+accumulated amounts of change.
+
+This makes sense in a continuous setting, but does not make sense in a discrete
+setting. In this respect Movements can be thought of as discretized Flows, or
+alternatively as the inverse of a Flow. Where a Flow might expressed "X units
+per second", a Movement will instead look at the seconds until a particular
+Movement is to occur. When aggregated by time, Movements approximate a Flow.
+
+### Models
+
+Models are "the rest" of the code. Typically these own Stocks, wire dependencies
+and potentially maintain other state.
+
+In Skenario the main toplevel Models are the KnativeAutoscaler and the Cluster.
+These establish the various initial Stocks (eg. RequestsBuffered, ReplicasLaunching)
+that are used during the life of the simulation.
+
+Generally speaking, keep logic out of Models. Everything that can be expressed
+as a Movement ought to be; special-case logic should be placed in `Add()` or
+`Remove()`.
+
+### Environment
+
+The Environment is the effective root of the program. Its `Run()` method has the
+core simulation loop and the `AddToSchedule()` method provides the means for
+simulation components to set up future Movements. The Environment also provides a
+number of accessors to key global variables: the simulation's `CurrentMovementTime()`,
+the simulation's `HaltTime()` and a `Context()` that holds a logger object as a value.
+
+The Environment schedules two specialised Movements: the `start_to_running` Movement
+and the `running_to_halted` Movement. These are placed at the boundary points of time.
+That is: only the start Movement may occur at time zero, only the halt Movement may
+occur at the halting time. Other Movements must occur between these two, meaning that
+the permissible time range is expressible as `(0, halt)` or as `[1, halt-1]`.
+
+## How a scenario is executed
+
+### `Run()`
+
+The Environment's `Run()` method implements the "next-event list" concept from
+Discrete Event Simulation. This means that on each iteration, the Environment will
+select the next Movement from a queue and execute it.
+
+Ordering is by the `OccursAt` time of Movements. Internally, the Environment is
+relying on a `MovementPriorityQueue` to maintain orderly records; ultimately this is
+a basic heap.
+
+Once a Movement has been dequeued, the simulation's current time is advanced to
+the `OccursAt` value of the Movement. The Environment then calls the `Remove()` method
+of the `From()` Stock to retrieve an Entity. If that call is successful, it then
+`Add()`s that Entity to the `To()` stock.
+
+On each iteration, Movements that occurred without error are captured in the
```

Comment:
What is the meaning of an error? Does it model an error in the simulated system?